tableaudocumentapi/folder.py

```diff
@@ -0,0 +1,171 @@
+from functools import wraps
+import xml.etree.ElementTree as ET
+
+from tableaudocumentapi import Field
+
+########
+# This is needed in order to determine if something is a string or not.  It is necessary because
+# of differences between python2 (basestring) and python3 (str).  If python2 support is every
+# dropped, remove this and change the basestring references below to str
+try:
+    basestring
+except NameError:  # pragma: no cover
+    basestring = str
+
+
+class AlreadyMemberOfThisFolderException(Exception):
+    pass
+
+
+class MemberOfMultipleFoldersException(Exception):
+    pass
+
+
+def argument_is_one_of(*allowed_values):
+    def property_type_decorator(func):
+        @wraps(func)
+        def wrapper(self, value):
+            if value not in allowed_values:
+                error = "Invalid argument: {0}. {1} must be one of {2}."
+                msg = error.format(value, func.__name__, allowed_values)
+                raise ValueError(error)
+            return func(self, value)
+        return wrapper
+    return property_type_decorator
+
+
+class FolderItem(object):
+    """ FolderItems belong to Folders and describe the Field-Objects
+    that belong to a folder
+    """
+
+    def __init__(self, name, _type):
+        self.name = name
+        self.type = _type
+
+    @classmethod
+    def from_xml(cls, xml):
+        return cls(xml.get('name'), xml.get('type'))
+
+    @classmethod
+    def from_field(cls, field):
+        return cls(field.id, 'field')
+
+
+class Folder(object):
+    """ This class represents a folder in a Datasource.
+
+    Folders have a name, a role (dimensions or measures) and contain Items
+    """
+
+    def __init__(self, datasource, xml):
+        self._datasource = datasource
+        self._xml = xml
+        self.name, self.role = self._xml.get('name'), self._xml.get('role')
```

Comment:
Separate lines for each item (easier to scan) please.

And I believe `.get` returns `None` if there's no match already, but we try to remember to have it explicitly as `.get('thing', None)` for readability